### PR TITLE
Expose API for reservoir manipulation

### PR DESCRIPTION
**Pull request by pfmooney**
This includes a rust clone of `rsrvrctl` as an example of the API usage.

### CODE REVIEW

File: bin/propolis-utils/src/bin/rsrvrctl.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        Command::Set { sz, chunk } => {
+            do_resize(&ctl, sz * MB, chunk.unwrap_or(0))?;
+        }
+        Command::Query => {
+            let sz = ctl.reservoir_query()?;
+            println!(
+                "Free KiB:\t\t\t{}
+Allocated KiB:\t\t\t{}
+Transient Allocated KiB:\t{}
+Size limit KiB:\t\t\t{}",
```
Comment (luqmana) [inline bin/propolis-utils/src/bin/rsrvrctl.rs:86]: You can line this up in the source too by adding `\` at the end to ignore any leading whitespace on the next line.
```suggestion
                "Free KiB:\t\t\t{}\n\
                 Allocated KiB:\t\t\t{}\n\
                 Transient Allocated KiB:\t{}\n\
                 Size limit KiB:\t\t\t{}",
```

File: bin/propolis-utils/src/bin/rsrvrctl.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    let ctl = VmmCtlFd::open()?;
+
+    let size_total =
+        |q: bhyve_api::vmm_resv_query| q.vrq_free_sz + q.vrq_alloc_sz;
+
+    match opts.cmd {
+        Command::Add { sz, chunk } => {
+            let cur_sz = ctl.reservoir_query().map(size_total)?;
+            do_resize(&ctl, cur_sz + sz * MB, chunk.unwrap_or(0))?;
```
Comment (luqmana) [inline bin/propolis-utils/src/bin/rsrvrctl.rs:66]: should this be a `saturating_add` to match the remove case?

Comment (pfmooney) [reply bin/propolis-utils/src/bin/rsrvrctl.rs]: Yeah, good catch.

**Comment by smklein** [comment]
FYI, I'm still working through it, but https://github.com/oxidecomputer/omicron/pull/2684 seems related

**Comment by pfmooney** [comment]
> FYI, I'm still working through it, but [oxidecomputer/omicron#2684](https://github.com/oxidecomputer/omicron/pull/2684) seems related

Yeah, I tagged you here so you'd be aware of the API addition so it could be used elsewhere (like in oxidecomputer/omicron#2684)

**Comment by smklein** [comment]
> > FYI, I'm still working through it, but [oxidecomputer/omicron#2684](https://github.com/oxidecomputer/omicron/pull/2684) seems related
> 
> Yeah, I tagged you here so you'd be aware of the API addition so it could be used elsewhere (like in [oxidecomputer/omicron#2684](https://github.com/oxidecomputer/omicron/pull/2684))

Just to clarify:

- This PR adds a binary to `propolis-utils`, which, AFAICT, are developer tools which I should not depend on from Omicron.
- It also adds an interface around an ioctl in the `bhyve-api` crate, which we don't *currently* depend on directly from Omicron.

IMO this is fine; it's great for developers using Propolis, and I'm happy with this PR, but I was planning on just calling the `rsrvrctl` CLI tool directly during initialization from Omicron. Does that seem reasonable to you?

**Comment by pfmooney** [comment]
>     * This PR adds a binary to `propolis-utils`, which, AFAICT, are developer tools which I should not depend on from Omicron.
>     * It also adds an interface around an ioctl in the `bhyve-api` crate, which we don't _currently_ depend on directly from Omicron.

Correct, the `rsrvrctl` added to `propolis-utils` here wasn't meant to be used by Omicron, but rather serve as an example of how one might consume the API added to `bhyve-api`.


>  I was planning on just calling the `rsrvrctl` CLI tool directly during initialization from Omicron. Does that seem reasonable to you?

The interface offered by the in-gate `rsrvrctl` is not Committed or anything (although neither is the ioctl API exposed via `bhyve-api`).  I added the interface to those ioctls so they could be used by sled-agent for this task.  It seems to me like an easier way of achieving it, and being able to handle any errors along the way.

**Pull request merged**